### PR TITLE
Wireshark Dissector for Netwayste

### PR DESCRIPTION
**Pull request by Manghi**
- [x] Do the hard part and get the foundations written (GJ Aaron!)
- [x] Update README and build instructions
- [x] Limit `unsafe` usage and scope
- [x] Parse AST of `netwayste/src/net.rs`
- [x] Generate header fields ast AST
- [x] Dissect packet using AST
- [x] Clean up dissection view by adding sub-trees for lists, where applicable
- [x] Unit test parser
- [x] Attend to `PR_GATE` notices
- [x] Attend to `FIXME` notices
- [x] Review `TODO` notices
- [x] Add documentation
- [x] No warnings

### CODE REVIEW

**Comment by Manghi** [comment]
Latest commit seems to have broken my build. It doesn't seem to find `/usr/include/wireshark/` which is where `config.h` exists on my system. Here's what it does find after your update to traverse the include paths and printing the `linker_arg`.

```
-I/usr/local/include/wireshark
-I/usr/include/glib-2.0
-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
-I/usr/include
```

Still looking into why.

Full log below:

```
mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$ cargo build
   Compiling dissect-netwayste v0.1.0 (/home/mang/Code/conwayste/dissect-netwayste)
error: failed to run custom build command for `dissect-netwayste v0.1.0 (/home/mang/Code/conwayste/dissect-netwayste)`

Caused by:
  process didn't exit successfully: `/home/mang/Code/conwayste/target/debug/build/dissect-netwayste-a77332e9af447255/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
cargo:rustc-link-lib=glib-2.0
cargo:rustc-link-search=native=/usr/local/lib
cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
cargo:rustc-link-lib=wireshark
cargo:rustc-link-lib=gmodule-2.0
cargo:rustc-link-lib=gthread-2.0
cargo:rustc-link-lib=glib-2.0
cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
cargo:rustc-link-lib=glib-2.0
cargo:rustc-link-search=native=/usr/local/lib
cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
cargo:rustc-link-lib=wireshark
cargo:rustc-link-lib=gmodule-2.0
cargo:rustc-link-lib=gthread-2.0
cargo:rustc-link-lib=glib-2.0

--- stderr
wrapper.h:2:10: fatal error: 'config.h' file not found
wrapper.h:2:10: fatal error: 'config.h' file not found, err: true
thread 'main' panicked at 'Unable to generate bindings: ()', dissect-netwayste/build.rs:22:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$ whereis wireshark
wireshark: /usr/bin/wireshark /usr/lib/x86_64-linux-gnu/wireshark /etc/wireshark /usr/local/bin/wireshark /usr/local/lib/wireshark /usr/include/wireshark /usr/share/wireshark /usr/share/man/man1/wireshark.1.gz
mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$ ldconfig -p | grep wireshark
	libwireshark.so.13 (libc6,x86-64) => /usr/local/lib/libwireshark.so.13
	libwireshark.so.13 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libwireshark.so.13
	libwireshark.so (libc6,x86-64) => /usr/local/lib/libwireshark.so
	libwireshark.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libwireshark.so
```

**Comment by AaronM04** [comment]
What is the output of these commands?

pkg-config glib-2.0
pkg-config wireshark

On April 12, 2020 9:30:12 PM PDT, Manghi <notifications@github.com> wrote:
>Latest commit seems to have broken my build. It doesn't seem to find
>`/usr/include/wireshark/` which is where `config.h` exists on my
>system. Here's what it does find after your update to traverse the
>include paths and printing the `linker_arg`.
>
>```
>-I/usr/local/include/wireshark
>-I/usr/include/glib-2.0
>-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
>-I/usr/include
>```
>
>Still looking into why.
>
>Full log below:
>
>```
>mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$ cargo
>build
>Compiling dissect-netwayste v0.1.0
>(/home/mang/Code/conwayste/dissect-netwayste)
>error: failed to run custom build command for `dissect-netwayste v0.1.0
>(/home/mang/Code/conwayste/dissect-netwayste)`
>
>Caused by:
>process didn't exit successfully:
>`/home/mang/Code/conwayste/target/debug/build/dissect-netwayste-a77332e9af447255/build-script-build`
>(exit code: 101)
>--- stdout
>cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
>cargo:rustc-link-lib=glib-2.0
>cargo:rustc-link-search=native=/usr/local/lib
>cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
>cargo:rustc-link-lib=wireshark
>cargo:rustc-link-lib=gmodule-2.0
>cargo:rustc-link-lib=gthread-2.0
>cargo:rustc-link-lib=glib-2.0
>cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
>cargo:rustc-link-lib=glib-2.0
>cargo:rustc-link-search=native=/usr/local/lib
>cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu
>cargo:rustc-link-lib=wireshark
>cargo:rustc-link-lib=gmodule-2.0
>cargo:rustc-link-lib=gthread-2.0
>cargo:rustc-link-lib=glib-2.0
>
>--- stderr
>wrapper.h:2:10: fatal error: 'config.h' file not found
>wrapper.h:2:10: fatal error: 'config.h' file not found, err: true
>thread 'main' panicked at 'Unable to generate bindings: ()',
>dissect-netwayste/build.rs:22:20
>note: run with `RUST_BACKTRACE=1` environment variable to display a
>backtrace
>
>mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$
>whereis wireshark
>wireshark: /usr/bin/wireshark /usr/lib/x86_64-linux-gnu/wireshark
>/etc/wireshark /usr/local/bin/wireshark /usr/local/lib/wireshark
>/usr/include/wireshark /usr/share/wireshark
>/usr/share/man/man1/wireshark.1.gz
>mang@wayste:~/Code/conwayste/dissect-netwayste (aaron/dissector)$
>ldconfig -p | grep wireshark
>	libwireshark.so.13 (libc6,x86-64) => /usr/local/lib/libwireshark.so.13
>	libwireshark.so.13 (libc6,x86-64) =>
>/usr/lib/x86_64-linux-gnu/libwireshark.so.13
>	libwireshark.so (libc6,x86-64) => /usr/local/lib/libwireshark.so
>	libwireshark.so (libc6,x86-64) =>
>/usr/lib/x86_64-linux-gnu/libwireshark.so
>```
>
>
>-- 
>You are receiving this because your review was requested.
>Reply to this email directly or view it on GitHub:
>https://github.com/conwayste/conwayste/pull/103#issuecomment-612744704

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Manghi** [comment]
> What is the output of these commands? 
> pkg-config glib-2.0 
> pkg-config wireshark

Output is blank. The `--cflags` arg mirrors what was found during the build script output above.

```
mang@wayste:~ $ pkg-config wireshark
mang@wayste:~ $ pkg-config glib-2.0
mang@wayste:~ $ pkg-config --cflags wireshark
-pthread -I/usr/local/include/wireshark -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include
mang@wayste:~ $ pkg-config --cflags glib-2.0
-I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include
```

**Comment by AaronM04** [comment]
I got a weird panic when trying to dissect this file: https://iforgotmy.name/data/bc97a31cea232df852c9b0a6a6de267b/problem2020-04-16.pcapng

I added logging for the two variables involved (`discriminant` and `variants`) and got this:
```
discriminant is 1735876964 and variants is ["GameStart", "NewUserList", "GameFinish", "GameClose"]
```

The panic's stacktrace:
```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', dissect-netwayste/src/lib.rs:211:41
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:476
  11: rust_begin_unwind
             at src/libstd/panicking.rs:380
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  13: core::panicking::panic
             at src/libcore/panicking.rs:52
  14: core::option::Option<T>::unwrap
             at /rustc/b8cedc00407a4c56a3bda1ed605c6fc166655447/src/libcore/macros/mod.rs:10
  15: dissect_netwayste::ConwaysteTree::decode_nw_data_format
             at dissect-netwayste/src/lib.rs:211
  16: dissect_netwayste::ConwaysteTree::add_field_to_tree
             at dissect-netwayste/src/lib.rs:308
  17: dissect_netwayste::ConwaysteTree::decode_nw_data_format
             at dissect-netwayste/src/lib.rs:227
  18: dissect_netwayste::ConwaysteTree::add_field_to_tree
             at dissect-netwayste/src/lib.rs:308
  19: dissect_netwayste::ConwaysteTree::decode_nw_data_format
             at dissect-netwayste/src/lib.rs:222
  20: dissect_netwayste::ConwaysteTree::decode
             at dissect-netwayste/src/lib.rs:198
  21: dissect_netwayste::dissect_conwayste
             at dissect-netwayste/src/lib.rs:375
  22: <unknown>
  23: <unknown>
  24: dissector_try_uint_new
  25: dissector_try_uint
  26: decode_udp_ports
  27: <unknown>
  28: <unknown>
  29: <unknown>
  30: <unknown>
  31: dissector_try_uint_new
  32: <unknown>
  33: <unknown>
  34: <unknown>
  35: <unknown>
  36: dissector_try_uint_new
  37: dissector_try_uint
  38: <unknown>
  39: <unknown>
  40: <unknown>
  41: call_dissector_with_data
  42: <unknown>
  43: <unknown>
  44: <unknown>
  45: <unknown>
  46: <unknown>
  47: <unknown>
  48: <unknown>
  49: call_dissector_with_data
  50: <unknown>
  51: epan_dissect_run_with_taps
  52: <unknown>
  53: <unknown>
  54: <unknown>
  55: <unknown>
  56: main
  57: __libc_start_main
  58: <unknown>
```

**Comment by AaronM04** [comment]
I updated the cross Docker image, but the Travis build was failing due to a bash script error. I fixed that. Let's see if it works...

**Comment by Manghi** [comment]
> I got a weird panic when trying to dissect this file: 
> ...

Should work now! I was missing the case when nested structures (like `BroadcastChatMessage`) are part of lists. I've held off fully-fleshing out lists as I'd like to get them to appear in the dissection view drop hierarchy. Going to work on this next :)

**Comment by AaronM04** [comment]
I can confirm -- there is no panic now. GJ!

**Comment by AaronM04** [comment]
@Manghi you can merge from master in this branch now.

File: dissect-netwayste/src/netwaysteparser.rs [hunk 1]
```diff
@@ -282,3 +290,321 @@ pub fn parse_netwayste_format() -> HashMap<CString, NetwaysteDataFormat> {
...
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_parse_size_from_type_for_u8() {
+        let string = "u8".to_owned();
+        let expected_sizing = vec![Sizing::Fixed(1)];
+
+        let parsed_sizing = parse_size_from_type(string);
```
Comment (AaronM04) [inline dissect-netwayste/src/netwaysteparser.rs:303]: I like the test coverage in this file!

Can you please factor out this comparison code into a function though?

File: dissect-netwayste/src/netwaysteparser.rs [hunk 2]
```diff
@@ -108,32 +108,31 @@ fn parse_size_from_type(type_arg: String) -> Vec<Sizing> {
         }
 
         // Consume characters until we reach the inner type
-        let mut characters = param.chars();
```
Comment (AaronM04) [inline dissect-netwayste/src/netwaysteparser.rs]: This is really a comment about the first line in this function: if `param` is what you want to call this, you could just have that in the function arguments, then there's no need for `let param = type_arg;`

**Comment by AaronM04** [comment]
Can this be checked now?
>  Clean up dissection view by adding sub-trees for lists, where applicable

**Comment by Manghi** [comment]
> Can this be checked now?
> 
> > Clean up dissection view by adding sub-trees for lists, where applicable

There's a bug I've observed after merging in master that I need to first fix. It's definitely in my code and I have a good idea as to where it is. It needed the master.

File: dissect-netwayste/src/netwaysteparser.rs (hunk 2 above)
Comment (Manghi) [reply dissect-netwayste/src/netwaysteparser.rs]: Ah thanks! I think initially I wanted to preserve the `type_arg` for some reason but I moved away from it.

File: dissect-netwayste/src/netwaysteparser.rs (hunk 1 above)
Comment (Manghi) [reply dissect-netwayste/src/netwaysteparser.rs:303]: Aye!

**Comment by AaronM04** [comment]
Looks great now!

**Pull request merged**